File: plugins/resolve/aksharo_core_app/session.py
"""Sign-in/session state mirrored by the C09 Studio panel from this script's
own bridge device-code flow (B08b, `bridge/device_auth.py`).

The panel has no Resolve API of its own (brief: "the panel talks only to
C08's in-Resolve loopback server"), so it cannot run the device-code flow
itself; instead this script owns one `SessionState` and exposes it read-only
over `session.status` (`server.py`'s `PanelDeps`). The script updates the
state as `device_auth.bootstrap_device_credentials`'s callbacks fire
(pairing code shown, then signed in) — that wiring is not part of this
module, which only holds the state and its wire shape.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class SessionState:
    signed_in: bool = False
    workspace_id: str | None = None
    user_email: str | None = None
    pairing_user_code: str | None = None
    pairing_verification_url: str | None = None

    def to_wire(self) -> dict[str, object]:
        return {
            "signedIn": self.signed_in,
            "workspaceId": self.workspace_id,
            "userEmail": self.user_email,
            "pairing": (
                {
                    "userCode": self.pairing_user_code,
                    "verificationUrl": self.pairing_verification_url,
                }
                if self.pairing_user_code is not None
                else None
            ),
        }

    def mark_pairing(self, user_code: str, verification_url: str) -> None:
        """A device code was issued and is awaiting approval in the system
        browser (`device_auth.PairingCodeInfo`)."""
        self.pairing_user_code = user_code
        self.pairing_verification_url = verification_url

    def mark_signed_in(self, workspace_id: str, user_email: str) -> None:
        self.signed_in = True
        self.workspace_id = workspace_id
        self.user_email = user_email
        self.pairing_user_code = None
        self.pairing_verification_url = None

    def mark_signed_out(self) -> None:
        self.signed_in = False
        self.workspace_id = None
        self.user_email = None
        self.pairing_user_code = None
        self.pairing_verification_url = None


__all__ = ["SessionState"]

File: plugins/resolve-panel/src/auth/session.ts
/**
 * Sign-in state mirrored from the script's own bridge session (brief item 1: "sign-in state
 * mirrored from the script"), not a device-code flow this panel runs itself — the script
 * (`aksharo_core_app/bridge/device_auth.py`, B08b) is the one process that talks to the
 * API's device-code endpoints; this panel only polls `session.status`
 * (`aksharo_core_app/session.py`'s wire shape) and renders it.
 *
 * THREAT-MODEL.md T13 / D25 ("panels keep tokens in memory only") does not even apply here —
 * this panel never holds a token at all, unlike C05a's UXP panel — but the same "no
 * persistence" discipline is kept for the mirrored state below.
 */

export interface PairingInfo {
  readonly userCode: string;
  readonly verificationUrl: string;
}

export interface SessionStatus {
  readonly signedIn: boolean;
  readonly workspaceId: string | null;
  readonly userEmail: string | null;
  readonly pairing: PairingInfo | null;
}

/** `session.status`'s raw JSON-RPC result shape (`session.py`'s `SessionState.to_wire()`). */
export interface SessionStatusWire {
  readonly signedIn: boolean;
  readonly workspaceId: string | null;
  readonly userEmail: string | null;
  readonly pairing: { readonly userCode: string; readonly verificationUrl: string } | null;
}

export function parseSessionStatus(wire: SessionStatusWire): SessionStatus {
  return {
    signedIn: wire.signedIn,
    workspaceId: wire.workspaceId,
    userEmail: wire.userEmail,
    pairing: wire.pairing
      ? { userCode: wire.pairing.userCode, verificationUrl: wire.pairing.verificationUrl }
      : null,
  };
}

export const SIGNED_OUT_STATUS: SessionStatus = {
  signedIn: false,
  workspaceId: null,
  userEmail: null,
  pairing: null,
};
